app/src/main/java/ch/epfl/cs311/wanderwave/viewmodel/MapViewModel.kt

```diff
@@ -103,18 +127,59 @@ constructor(val locationSource: LocationSource, private val beaconRepository: Be
    * @author Menzo Bouaissi
    * @since 4.0
    */
-  fun getRandomSong(id: String) {
+  fun getRandomSongAndAddToProfile(beaconId: String) {
     viewModelScope.launch {
-      beaconRepository.getItem(id).collect { fetchedBeacon ->
+      beaconRepository.getItem(beaconId).collect { fetchedBeacon ->
         fetchedBeacon.onSuccess { beacon ->
-          if (beacon.profileAndTrack.isEmpty()) {
-            Log.e("No songs found", "No songs found for the given id")
-            return@onSuccess
+          if (beacon.profileAndTrack.isNotEmpty()) {
+            _retrievedSongs.value = beacon.profileAndTrack.random()
+            trackRepository.addItemsIfNotExist(listOf(_retrievedSongs.value.track))
           }
-          _retrievedSongs.value = beacon.profileAndTrack.random()
         }
         fetchedBeacon.onFailure { exception ->
-          Log.e("No beacons found", "No beacons found for the given id")
+          Log.e("No beacons found", "No beacons found for the given id $exception")
+        }
+      }
+    }
+  }
+
+  fun retrieveSongFromProfileAndAddToBeacon(beaconId: String) {
```

Comment:
Maybe make the fact that the song is chosen randomly reflect in the name of this function.